### ISSUE TITLE
Not able to reproduce the results in the MTEB benchmark for qwen-3 0.6B embedding model

### ISSUE THREAD

**Issue by rkoystart** [embeddings-benchmark/mteb#2867]
### Describe the bug

I have tried to reproduce the results mentioned in the MTEB benchmark for the qwen-3 0.6 embedding model

### To reproduce

```
import mteb

def eval():
        st_model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B", trust_remote_code=True)
        task_dataset_name = "SCIDOCS"
	tasks = mteb.get_tasks(tasks=[task_dataset_name])
	evaluation = mteb.MTEB(tasks=tasks)
	results = evaluation.run(st_model, output_folder=f"{task_dataset_name}_prompts", eval_splits = ["test"], batch_size=1, encode_kwargs={"normalize_embeddings":False})


```

When i run the above function i am not getting the value present in the MTEB benchmark.
Value Present in the MTEB benchmark = 24.41 (ndcg_at_10)
Value i am getting while running the above file = 22.75 (ndcg_at_10)

The sentence transformer version is 4.1.0
the MTEB version is 1.38.29
I am facing the above issue not only for SCIDOCS dataset, but also for may other dataset as well.
GPU is A6000

<img width="1505" alt="Image" src="https://github.com/user-attachments/assets/6b9ec22a-16c7-479d-b8f9-a2a86ff00f00" />

### Are you interested to contribute a fix for this bug?

No

**Comment by Samoed** [comment]
To have correct results, you should load model with 
```python
import mteb
model = mteb.get_model("Qwen/Qwen3-Embedding-0.6B")
...
```

**Comment by rkoystart** [comment]
@Samoed thanks for the immediate response. 
But could i know what is the difference between the way you have created the model instance and the way i have created the model instance using sentence transformers library.
I thought creating the model instance in any of the above way would give the same results.

**Comment by Samoed** [comment]
Yes, underneath we're using `SentenceTransformers`, but we have a wrapper that applies prompts to the texts, which can lead to differences in performance

**Comment by rkoystart** [comment]
@Samoed i tried to run the above mentioned code using the model instance created from `mteb.get_model("model_name")`. For SCIDOCS i am able to get the almost same metrics for most of the datasets. I have the following queries 
1)  But for the dataset `TRECCOVID` i am getting the scores as 89.30. But the score mentioned in the leaderboard is 90.52. Am i missing out anything.

2) Also while using the sentence transformers model to make sure the prompts are being applied properly i have made the below changes just so that the code works for the Qwen-3 model. (Not sure whether this is the right approach).
In SentenceTransformerWrapper
```
class SentenceTransformerWrapper(Wrapper):
    def __init__(
        self,
        model: str | SentenceTransformer | CrossEncoder,
        revision: str | None = None,
        model_prompts: dict[str, str] | None = None,
        **kwargs,
    ) -> None:
        """Wrapper for SentenceTransformer models.

        Args:
            model: The SentenceTransformer model to use. Can be a string (model name), a SentenceTransformer model, or a CrossEncoder model.
            revision: The revision of the model to use.
            model_prompts: A dictionary mapping task names to prompt names.
                First priority is given to the composed prompt of task name + prompt type (query or passage), then to the specific task prompt,
                then to the composed prompt of task type + prompt type, then to the specific task type prompt,
                and finally to the specific prompt type.
            **kwargs: Additional arguments to pass to the SentenceTransformer model.
        """
        if isinstance(model, str):
            self.model = SentenceTransformer(model, revision=revision, **kwargs)
        else:
            self.model = model

        if (
            model_prompts is None
            and hasattr(self.model, "prompts")
            and len(self.model.prompts) > 0
        ):
            try:
                # model_prompts = self.validate_task_to_prompt_name(self.model.prompts) commented this line and added the below line
                model_prompts = self.model.prompts
            except KeyError:
                model_prompts = None
                logger.warning(
                    "Model prompts are not in the expected format. Ignoring them."
                )
        elif model_prompts is not None and hasattr(self.model, "prompts"):
            logger.info(f"Model prompts will be overwritten with {model_prompts}")
            self.model.prompts = model_prompts
        # self.model_prompts = self.validate_task_to_prompt_name(model_prompts). commented this line and added the below line
        self.model_prompts = self.model.prompts

        if isinstance(self.model, CrossEncoder):
            self.predict = self._predict

        if hasattr(self.model, "similarity") and callable(self.model.similarity):
            self.similarity = self.model.similarity

```
Would like to know your suggestion for the above changes.

**Comment by Samoed** [comment]
`Qwen3-embeddings` uses `InstructSentenceTransformerWrapper`, not `SentenceTransformerWrapper`. Can you give full code of your run for reproduction?

**Comment by rkoystart** [comment]
Yeah, just check that the Qwen-3 model used `InstructSentenceTransformerWrapper` not `SentenceTransformerWrapper`. when the model is created using `mteb.get_model()`

I have 2 cases where 1) i use the model from the `SentenceTransformer` library 2) create the model instance using the `mteb.get_model()`

For use case 1, the `SentenceTransformer` model is being passed to `SentenceTransformerWrapper` and used further. I have modified the `SentenceTransformerWrapper` as below (so that model related prompts are applied)

```
from __future__ import annotations

import logging
from collections.abc import Sequence
from typing import Any

import numpy as np
import torch
from sentence_transformers import CrossEncoder, SentenceTransformer

from mteb.encoder_interface import PromptType
from mteb.models.wrapper import Wrapper

logger = logging.getLogger(__name__)


class SentenceTransformerWrapper(Wrapper):
    def __init__(
        self,
        model: str | SentenceTransformer | CrossEncoder,
        revision: str | None = None,
        model_prompts: dict[str, str] | None = None,
        **kwargs,
    ) -> None:
        """Wrapper for SentenceTransformer models.

        Args:
            model: The SentenceTransformer model to use. Can be a string (model name), a SentenceTransformer model, or a CrossEncoder model.
            revision: The revision of the model to use.
            model_prompts: A dictionary mapping task names to prompt names.
                First priority is given to the composed prompt of task name + prompt type (query or passage), then to the specific task prompt,
                then to the composed prompt of task type + prompt type, then to the specific task type prompt,
                and finally to the specific prompt type.
            **kwargs: Additional arguments to pass to the SentenceTransformer model.
        """
        if isinstance(model, str):
            self.model = SentenceTransformer(model, revision=revision, **kwargs)
        else:
            self.model = model

        if (
            model_prompts is None
            and hasattr(self.model, "prompts")
            and len(self.model.prompts) > 0
        ):
            try:
                # model_prompts = self.validate_task_to_prompt_name(self.model.prompts)
                model_prompts = self.model.prompts.  # commented the previous line and added this line
            except KeyError:
                model_prompts = None
                logger.warning(
                    "Model prompts are not in the expected format. Ignoring them."
                )
        elif model_prompts is not None and hasattr(self.model, "prompts"):
            logger.info(f"Model prompts will be overwritten with {model_prompts}")
            self.model.prompts = model_prompts # commented the next line and added this line
        # self.model_prompts = self.validate_task_to_prompt_name(model_prompts)
        self.model_prompts = self.model.prompts

        if isinstance(self.model, CrossEncoder):
            self.predict = self._predict

        if hasattr(self.model, "similarity") and callable(self.model.similarity):
            self.similarity = self.model.similarity

    def encode(
        self,
        sentences: Sequence[str],
        *,
        task_name: str,
        prompt_type: PromptType | None = None,
        **kwargs: Any,
    ) -> np.ndarray:
        """Encodes the given sentences using the encoder.

        Args:
            sentences: The sentences to encode.
            task_name: The name of the task. Sentence-transformers uses this to
                determine which prompt to use from a specified dictionary.
            prompt_type: The name type of prompt. (query or passage)
            **kwargs: Additional arguments to pass to the encoder.

            The order of priorities for prompt selection are:
                1. Composed prompt of task name + prompt type (query or passage)
                2. Specific task prompt
                3. Composed prompt of task type + prompt type (query or passage)
                4. Specific task type prompt
                5. Specific prompt type (query or passage)


        Returns:
            The encoded sentences.
        """
        prompt_name = None
        if self.model_prompts is not None:
            prompt_name = self.get_prompt_name(
                self.model_prompts, task_name, prompt_type
            )
        if prompt_name:
            logger.info(
                f"Using prompt_name={prompt_name} for task={task_name} prompt_type={prompt_type}"
            )
        else:
            logger.info(
                f"No model prompts found for task={task_name} prompt_type={prompt_type}"
            )
        logger.info(f"Encoding {len(sentences)} sentences.")
        print(" line sentence transformer wrapper line number 106 and prompt is ", prompt_name, self.model_prompts)
        embeddings = self.model.encode(
            sentences,
            prompt_name=prompt_name,
            **kwargs,
        )
        if isinstance(embeddings, torch.Tensor):
            # sometimes in kwargs can be return_tensors=True
            embeddings = embeddings.cpu().detach().float().numpy()
        return embeddings

    def _predict(
        self,
        sentences: Sequence[str],
        **kwargs: Any,
    ) -> np.ndarray:
        return self.model.predict(
            sentences,
            convert_to_numpy=True,
            **kwargs,
        )

```
And the Inference code is use is
```
import mteb

def eval():
        st_model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B")
        st_model = mteb.get_model(model_name, trust_remote_code=True)
        task_dataset_name = "SCIDOCS"
	tasks = mteb.get_tasks(tasks=[task_dataset_name])
	evaluation = mteb.MTEB(tasks=tasks)
	results = evaluation.run(st_model, output_folder=f"{task_dataset_name}_prompts", eval_splits = ["test"], batch_size=1, encode_kwargs={"normalize_embeddings":False})

```
By doing this i see that the function that gets invoked for the embedding is in line number 109 in SentenceTransformerWrapper
```
embeddings = self.model.encode(
            sentences,
            prompt_name=prompt_name,
            **kwargs,
        )
```

What the above encode does for prompt_name = "query" is, applied the query related prompt to the given input sentences for Example
a) if the sentence is "hello world", the text gets converted to " {prompt related to the query } hello world"
If no prompt is present for the `prompt_name`, the inputs text gets encoded directly.

For Use case 2: Where the model is created using `mteb.get_model()` I see that instead of applying Model related prompt like above, i see task related prompts are being applied. For example for `SCIDOCS`, the text `hello world` is being converted as `Instruct: Given a scientific paper title, retrieve paper abstracts that are cited by the given paper hello world`. This is something related to the task and not model related prompt for query.

below function gets called in the InstructSentenceTransformerWrapper for encoding the input sentences.
```
    def encode(
        self,
        sentences: Sequence[str],
        *,
        task_name: str,
        prompt_type: PromptType | None = None,
        **kwargs: Any,
    ) -> np.ndarray:
        if self.add_eos_token:
            sentences = [
                example + self.model.tokenizer.eos_token for example in sentences
            ]

        instruction = self.get_task_instruction(task_name, prompt_type)

        # to passage prompts won't be applied to passages
        if not self.apply_instruction_to_passages and prompt_type == PromptType.passage:
            instruction = None
            logger.info(
                f"No instruction used, because prompt type = {prompt_type.passage}"
            )

        if instruction:
            logger.info(f"Using instruction: '{instruction}' for task: '{task_name}'")

        embeddings = self.model.encode(
            sentences,
            prompt=instruction,
            **kwargs,
        )

        if isinstance(embeddings, torch.Tensor):
            # sometimes in kwargs can be return_tensors=True
            embeddings = embeddings.cpu().detach().float().numpy()
        return embeddings

```


Inference Code for Case 2:
```
def eval():
        st_model = mteb.get_model("Qwen/Qwen3-Embedding-0.6B")
        task_dataset_name = "SCIDOCS"
	tasks = mteb.get_tasks(tasks=[task_dataset_name])
	evaluation = mteb.MTEB(tasks=tasks)
	results = evaluation.run(st_model, output_folder=f"{task_dataset_name}_prompts", eval_splits = ["test"], batch_size=1, encode_kwargs={"normalize_embeddings":False})
```

So out the above 2, i see while using SentenceTransformer based, model related prompts are being applied, but while using the model created using `model.get_model()` task related prompts are being applied.

Please correct me if i am wrong.
My doubt is i feel always the model related prompt needs to be applied irrespective whether the task related prompts are being applied or not.
Please clarify @Samoed

**Comment by Samoed** [comment]
> My doubt is i feel always the model related prompt needs to be applied irrespective whether the task related prompts are being applied or not.

I don't understand what means model related prompts. Models can define any kind of prompts, and we need to validate them to apply them correctly, because `SentenceTransformers` just prepend these prompts to the input before encoding.

Your code in case2 looks correct, except you shouldn't pass `normalize_embeddings` to `encode_kwargs`, but I don't have many resources to check results for `TRECCOVID`

**Comment by rkoystart** [comment]
> I don't understand what means model related prompts.

Here model related prompts i Mean the prompts present in the variable `st_model.prompts` in the below code.

```
from sentence_transformers import SentenceTransformer
st_model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B", trust_remote_code=True)
st_model.prompts
```

Whereas by task related prompts i mean the prompts mentioned in the Metadata of the tasks 

```
from __future__ import annotations

from mteb.abstasks.TaskMetadata import TaskMetadata

from ....abstasks.AbsTaskRetrieval import AbsTaskRetrieval


class SCIDOCS(AbsTaskRetrieval):
    metadata = TaskMetadata(
        name="SCIDOCS",
        dataset={
            "path": "mteb/scidocs",
            "revision": "f8c2fcf00f625baaa80f62ec5bd9e1fff3b8ae88",
        },
        description=(
            "SciDocs, a new evaluation benchmark consisting of seven document-level tasks ranging from citation"
            + " prediction, to document classification and recommendation."
        ),
        reference="https://allenai.org/data/scidocs",
        type="Retrieval",
        category="s2p",
        modalities=["text"],
        eval_splits=["test"],
        eval_langs=["eng-Latn"],
        main_score="ndcg_at_10",
        date=None,
        domains=["Academic", "Written", "Non-fiction"],
        task_subtypes=[],
        license="cc-by-sa-4.0",
        annotations_creators=None,
        dialect=[],
        sample_creation="found",
        bibtex_citation=r"""
@inproceedings{specter2020cohan,
  author = {Arman Cohan and Sergey Feldman and Iz Beltagy and Doug Downey and Daniel S. Weld},
  booktitle = {ACL},
  title = {SPECTER: Document-level Representation Learning using Citation-informed Transformers},
  year = {2020},
}
""",
        prompt={
            "query": "Given a scientific paper title, retrieve paper abstracts that are cited by the given paper"
        },
    )

```

> Models can define any kind of prompts, and we need to validate them to apply them correctly, because SentenceTransformers just prepend these prompts to the input before encoding.

What you said is right, we only need to make sure whether the prompt is being applied properly. By model related prompt being applied while using SentenceTransformerModel and Task Related prompt being applied while using the `InstructSentenceTransformerWrapper` model (created by `mteb.get_model()`) is what making me confuse.

I Think in both the cases, the prompt present in the variable st_model.prompts needs to be applied, but it is not happening.

**Comment by Samoed** [comment]
We can't create mapping between `model.prompts` and our prompt name easily, because each model have its own namings and we won't be able to change them.  I don't why commenting out prompt validation would solve this problem.

> I Think in both the cases, the prompt present in the variable st_model.prompts needs to be applied, but it is not happening.

To apply these prompts, we need to pass prompt_name that exist in our prompts list. Prompts in `model.prompts` are just recommended prompts for models and can be passed directly (that we do).

**Comment by rkoystart** [comment]
@Samoed there seems to be a gap in what in am trying to convey and your understanding. Sorry. But can you confirm me the below things
1) What prompt is applied for encoding  a list of query sentences while using the `SentenceTransformer` model
2) What prompt is applied for encoding a list of query sentences while using the `InstructSentenceTransformerWrapper` model

Please assume i am asking for the dataset `SCIDOCS`
?

My answer on debugging is 
1) While using the code in the previous comment for the `SentenceTransformerWrapper`, the prompt being applied for query is `Instruct: Given a web search query, retrieve relevant passages that answer the query\nQuery:',`

If the unchanged code is used for `SentenceTransformerWrapper` class, no prompt is being used for the query and the document.

2) While using the `InstructSentenceTransformerWrapper`, the prompt applied for query is `Given a scientific paper title, retrieve paper abstracts that are cited by the given paper`

**Comment by Samoed** [comment]
> What prompt is applied for encoding a list of query sentences while using the SentenceTransformer model

Prompt that defined in `query` or `Retrieval-query` or `TaskName-query`. 

> What prompt is applied for encoding a list of query sentences while using the InstructSentenceTransformerWrapper model

Same as for SentenceTransformerWrapper

> While using the code in the previous comment for the SentenceTransformerWrapper, the prompt being applied for query is Instruct: Given a web search query, retrieve relevant passages that answer the query\nQuery:',

Yes, this is `query` prompt from `model.prompts`

>  While using the InstructSentenceTransformerWrapper, the prompt applied for query is Given a scientific paper title, retrieve paper abstracts that are cited by the given paper

No, it will be `Instruct: Given a scientific paper title, retrieve paper abstracts that are cited by the given paper
Query: `. From my run logs
```
Using instruction: 'Instruct: Given a scientific paper title, retrieve paper abstracts that are cited by the given paper
Query:' for task: 'SCIDOCS'
```

**Comment by rkoystart** [comment]
>Yes, this is query prompt from model.prompts

> No, it will be Instruct: Given a scientific paper title, retrieve paper abstracts that are cited by the given paper Query: . From my run logs

So my question is for the same dataset, the same model instantiated using different class applies different prompt. This was my doubt. Is this the right way ?

**Comment by Samoed** [comment]
Yes, this is a right way. We don't recommend passing `SentenceTransformer` class directly to our evaluator. Users shoud use `mteb.get_model`
